test/test_torch.py

```diff
@@ -16330,16 +16329,112 @@ def test_svd_tall_some_col_maj(self, device, dtype):
     def test_svd_tall_all_col_maj(self, device, dtype):
         self._test_svd_helper((5, 20), False, True, device, dtype)
 
+
+# Below are fixtures and functions for generating tests related to unary
+# "universal functions (ufuncs)", a class of elementwise unary operations
+# This class can be further divided into "normal" and "floating" unary
+# ufuncs. The former does not participate in type promotion, while the latter
+# promotes integral types, including bool, to the default scalar type.
+
+# Helpers to organize ufunc metadata and improve test readability
+_ufunc_meta = collections.namedtuple('_ufunc_meta',
+                                     'is_floating alt_impl complex_on')
+def ufunc_meta(is_floating=False, alt_impl=None, complex_on=('cpu', 'cuda')):
+    return _ufunc_meta(is_floating, alt_impl, complex_on)
+
+# Dict of unary ufuncs and their associated test metadata
+unary_ufuncs = {
+    'acos': ufunc_meta(is_floating=True),
+    'asin': ufunc_meta(is_floating=True),
+    'ceil': ufunc_meta(is_floating=True),
+    'expm1': ufunc_meta(is_floating=True, complex_on=()),
+    'floor': ufunc_meta(is_floating=True),
+    'log': ufunc_meta(is_floating=True),
+    'log10': ufunc_meta(is_floating=True),
+    'log1p': ufunc_meta(is_floating=True, complex_on=()),
+    'log2': ufunc_meta(is_floating=True),
+    'sin': ufunc_meta(is_floating=True),
+    'sinh': ufunc_meta(is_floating=True),
+    'sqrt': ufunc_meta(is_floating=True),
+    'trunc': ufunc_meta(is_floating=True),
+    'digamma': ufunc_meta(is_floating=True),
+    'rsqrt': ufunc_meta(is_floating=True, complex_on=('cpu')),
+    'sigmoid': ufunc_meta(is_floating=True, complex_on=('cpu')),
+    'atan': ufunc_meta(is_floating=True, complex_on=('cpu')),
+    'cos': ufunc_meta(is_floating=True, complex_on=('cpu')),
+    'cosh': ufunc_meta(is_floating=True, complex_on=('cpu')),
+    'exp': ufunc_meta(is_floating=True, complex_on=('cpu')),
+    'tan': ufunc_meta(is_floating=True, complex_on=('cpu')),
+    'tanh': ufunc_meta(is_floating=True, complex_on=('cpu')),
+    'erf': ufunc_meta(is_floating=True, complex_on=()),
+    'erfc': ufunc_meta(is_floating=True, complex_on=()),
+    'erfinv': ufunc_meta(is_floating=True, complex_on=()),
+    'lgamma': ufunc_meta(is_floating=True, complex_on=())
```

Comment:
Ops which use `complex_on=()` are those which do not support Complex Type (`torch.complex64`, `torch.complex128`) Tensors on CPU and CUDA devices. By default, `complex_on` is set to `('cpu', 'cuda')`.